internal/datastore/proxy/relationshipintegrity.go

```diff
@@ -0,0 +1,397 @@
+package proxy
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"fmt"
+	"hash"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/authzed/spicedb/pkg/datastore"
+	"github.com/authzed/spicedb/pkg/datastore/options"
+	corev1 "github.com/authzed/spicedb/pkg/proto/core/v1"
+	"github.com/authzed/spicedb/pkg/spiceerrors"
+	"github.com/authzed/spicedb/pkg/tuple"
+)
+
+// KeyConfig is a configuration for a key used to sign relationships.
+type KeyConfig struct {
+	ID        string
+	ExpiredAt *time.Time
+	Bytes     []byte
+}
+
+type hmacConfig struct {
+	keyID     string
+	expiredAt *time.Time
+	hmacPool  *sync.Pool
+}
+
+var (
+	alg         = sha256.New
+	versionByte = byte(0x01)
+)
+
+// NewRelationshipIntegrityProxy creates a new datastore proxy that ensures the integrity of
+// relationships by using HMACs to sign the data. The current key is used to sign new data,
+// and the expired keys are used to verify old data, if any.
+func NewRelationshipIntegrityProxy(ds datastore.Datastore, currentKey KeyConfig, expiredKeys []KeyConfig) (datastore.Datastore, error) {
+	currentKeyHMAC := hmacConfig{
+		keyID:     currentKey.ID,
+		expiredAt: currentKey.ExpiredAt,
+		hmacPool: &sync.Pool{
+			New: func() any { return hmac.New(alg, currentKey.Bytes) },
+		},
+	}
+
+	if currentKey.ExpiredAt != nil {
+		return nil, spiceerrors.MustBugf("current key cannot have an expiration")
+	}
+
+	keysByID := make(map[string]hmacConfig, len(expiredKeys)+1)
+	keysByID[currentKey.ID] = currentKeyHMAC
+
+	for _, key := range expiredKeys {
+		if key.ExpiredAt == nil {
+			return nil, spiceerrors.MustBugf("expired key missing expiration time")
+		}
+
+		if _, ok := keysByID[key.ID]; ok {
+			return nil, spiceerrors.MustBugf("found duplicate key ID: %s", key.ID)
+		}
+
+		keysByID[key.ID] = hmacConfig{
+			keyID:     key.ID,
+			expiredAt: key.ExpiredAt,
+			hmacPool: &sync.Pool{
+				New: func() any { return hmac.New(alg, key.Bytes) },
```

Comment:
it's probably worth benchmarking sha3 here, since you can use it as a MAC directly instead of through an HMAC

Reply:
I also learned recently about https://github.com/google/highwayhash (and its Go native impl, https://github.com/minio/highwayhash), theoretically 1 order of magnitude faster than SHA256 thanks to SIMD, but says:

> HighwayHash is not a general-purpose cryptographic hash function (such as Blake2b, SHA-3 or SHA-2) and should not be used if strong collision resistance is required.

Reply:
Tried both and SHA3 was slower locally while highway is ever so slightly faster

Reply:
Did some prelim tests and highwayhash is faster by about ~50% while SHA3 seemed to be slower.

However, I'm concerned about the lack of strong collision resistance. We'll need to investigate the implications of that.

Reply:
Updated branch with the highwayhash impl for testing

Reply:
The scenario this attempts to protect from is unauthorized modification of the state of the database. This requires sophisticated attacks and computing running for days to find a collision that could reveal the seed. If someone has access to the database and the machinery to run such an attack, the hash is not going to get in the way honestly, they probably have enough incentives to find other ways. I don't think remote attacks, with all the level of indirections, can give a clear signal to an attacker.

This is a decision the customer could themselves make and we could make it configurable via a flag. If the customer considers only collision-resistant hashes acceptable and are willing to accept the tradeoff in computational cost, then they can choose to do so. Highwayhash is by no means a "weak" hash, but one that makes the attack cost be sufficiently high to be unfeasible to most attackers.

>By contrast, 'strong' hashes such as SipHash or HighwayHash require infeasible attacker effort to find a hash collision (an expected 2^32 guesses of m per the birthday paradox) or recover the seed (2^63 requests). These security claims assume the seed is secret. It is reasonable to suppose s is initially unknown to attackers, e.g. generated on startup or even per-connection. 

Consider the authors describe how complicated an attack is:
>A timing attack by Wool/Bar-Yosef recovers 13-bit seeds by testing all 8K possibilities using millions of requests, which takes several days (even assuming unrealistic 150 us round-trip times). It appears infeasible to recover 64-bit seeds in this way.

Reply:
The context here was that (highwayhash isn't suitable for hmac)[https://github.com/minio/highwayhash/issues/13#issuecomment-528356795] (or at least we don't know that it is)